chore: {:focus} and {:split-here} doc and Verification debugging techniques (#1884)
dafny-lang/dafny: docs/DafnyRef/UserGuide.md

### 24.8.1. Verification debugging {#sec-verification-debugging}

Let's assume one assertion is failing ("assertion might not hold" or "postcondition might not hold"). What should you do next?

The following section is textual description of the animation below, which illustrates the principle of debugging an assertion by computing the weakest precondition:  
![weakestpreconditionDemo](https://user-images.githubusercontent.com/3601079/157976402-83fe4d37-8042-40fc-940f-bcfc235c7d2b.gif)

#### 24.8.1.1. Failing postconditions {#sec-failing-postconditions}
Let's look at an example of a failing postcondition.
```dafny
method FailingPostcondition(b: bool) returns (i: int)
  ensures 2 <= i
{
  var j := if !b then 3 else 1;
  if b {
    return j;
  }//^^^^^^^ a postcondition might not hold on this return path.
  i := 2;
}
```

> One first thing you can do is replace the sentence `return j;` by two sentences `i := j; return;` to better understand what is wrong:

Reviewer: ```suggestion
One first thing you can do is replace the statement `return j;` by two statements `i := j; return;` to better understand what is wrong:
```

This list is not exhaustive but can definitely be useful to provide the next step to figure out why Dafny could not prove an assertion.

 Failing assert           | Suggested rewriting
--------------------------|---------------------------------------
 <br>`x := Y;`<br>`assert P[x];` | `assert P[Y];`<br>`x := Y;`<br>`assert P[x];`
 <br>`if B {`<br>&nbsp;&nbsp;`  assert P;`<br>&nbsp;&nbsp;`  ...`<br>`}` | `assert B ==> P;`<br>`if B {`<br>&nbsp;&nbsp;`  assert P;`<br>&nbsp;&nbsp;`  ...`<br>`}`
 <br>`if B {`<br>&nbsp;&nbsp;`  ...`<br>`} else {`<br>&nbsp;&nbsp;`  assert P;`<br>&nbsp;&nbsp;`  ...`<br>`}` | `assert !B ==> P;`<br>`if B {`<br>&nbsp;&nbsp;`  ...`<br>`} else {`<br>&nbsp;&nbsp;`  assert P;`<br>&nbsp;&nbsp;`  ...`<br>`}`
 <br><br>`if X {`<br>&nbsp;&nbsp;`  ...`<br>`} else {`<br>&nbsp;&nbsp;`  ...`<br>`}`<br>`assert A;` | `if X {`<br>&nbsp;&nbsp;`  ...`<br>&nbsp;&nbsp;`  assert A;`<br>`} else {`<br>&nbsp;&nbsp;`  ...`<br>&nbsp;&nbsp;`  assert A;`<br>`}`<br>`assert A;`
 <br><br><br><br><br>`assert forall x :: Q(x);` | [`forall x`](#sec-forall-statement)<br>&nbsp;&nbsp;`  ensures Q(x)`<br>`{`<br>&nbsp;&nbsp;`  assert Q(x);`<br>`};`<br>` assert forall x :: Q(x);`

>  <br><br><br><br><br>`assert forall x :: P(x) ==> Q(x);` | [`forall x | P(x)`](#sec-forall-statement)<br>&nbsp;&nbsp;`  ensures Q(x)`<br>`{`<br>&nbsp;&nbsp;`  assert Q(x);`<br>`};`<br>` assert forall x :: P(x) ==> Q(x);`

Reviewer: The formatting on this line isn't quite right.

Author: It isn't quite right in GitHub markdown, but I tested it for the generated website and it works well.

 <br><br><br>`assert A ==> B;` | `if A {`<br>&nbsp;&nbsp;`  assert B;`<br>`};`<br>`assert A ==> B;`
 <br><br>`assert A && B;` | `assert A;`<br>`assert B;`<br>`assert A && B;`
 <br>`assert P(x);`<br>where `P` is an [`{:opaque}`](#sec-opaque) predicate | [`reveal P();`](#sec-reveal-statement)<br>`assert P(x);`<br><br>
 `assert P(x);`<br>where `P` is an [`{:opaque}`](#sec-opaque) predicate<br><br> | [`assert P(x) by {`](#sec-assert-statement)<br>[&nbsp;&nbsp;`  reveal P();`](#sec-reveal-statement)<br>`}`
 `assert P(x);`<br>where `P` is not an [`{:opaque}`](#sec-opaque) predicate with a lot of `&&` in its body and is assumed | Make `P` [`{:opaque}`](#sec-opaque) so that if it's assumed, it can be proven more easily. You can always [reveal](#sec-reveal-statement) it when needed.
 `ensures P ==> Q` on a lemma<br> | `requires P ensures Q` to avoid accidentally calling the lemma on inputs that do not satisfy `P`
 `seq(size, i => P)` | `seq(size, i requires 0 <= i < size => P);`
  <br><br>`assert forall x :: G(i) ==> R(i);` |  `assert G(i0);`<br>`assert R(i0);`<br>`assert forall i :: G(i) ==> R(i);` with a guess of the `i0` that makes the second assert to fail.

>   <br><br>`assert forall i | 0 < i <= m :: P(i);` |  `assert forall i | 0 < i < m :: P(i);`<br>`assert forall i | i == m :: P(i);`<br>`assert forall i | 0 < i <= m :: P(i);`<br><br>

Reviewer: This formatting is also a bit off.

>   <br><br>`assert forall i | i == m :: P(m);` |  `assert P(m);`<br>`assert forall i | i == m :: P(i);`

Reviewer: As is this.

* A consecutive pair of lines in a [`calc` statement](#sec-calc-statement) forms _an assertion_.
* Every function or method call with a [`requires` clause](#sec-requires-clause) yields _one assertion per requires clause_[^precision-requires-clause]
  (special cases such as sequence indexing come with a special require clause that the index is within bounds).
* Assignments `o.f := E;` yield an _assertion_ that `o.f` is allowed by the enclosing [`modifies` clause](#sec-loop-framing).
* [Assign-such-that operators](#sec-update-and-call-statement) `x :| P(x)` yield an _assertion_ that `exists x :: P(x)`.
* Every [`ensures` clause](#sec-ensures-clause) yields an _assertion_ at the end of the method and on every return.

[^precision-requires-clause]: Dafny actually breaks things down further. For example, a precondition `requires A && B` or an assert statement `assert A && B;` turns into two assertions, more or less like `requires A requires B` and `assert A; assert B;`.

It is useful to mentally visualize all these assertions as a list (an _assertion batch_) that roughly follows the order in the code[^complexity-path-encoding],

> except for `ensures` assertions that appear before in the code but, for verification purposes, would appear at the end.

Reviewer: ```suggestion
except for `ensures` assertions that appear earlier in the code but, for verification purposes, would appear at the end.
```

Thus, to prove or disprove a single assertion within this list, all the assumptions Dafny needs are 1) the global context,

> and 2) every preceding assumptions (and previous assertions transformed in assumptions).

Reviewer: ```suggestion
and 2) every preceding assumption (and previous assertions transformed in assumptions).
```

## 24.8. Verification

In this section, we suggest a methodology to figure out [why a single assertion might not hold](#sec-verification-debugging), we explain how to [make the verifier more focused on particular assertions](#sec-assertion-batches) to devote more time to prove them, and we also give some more examples of [useful options and attributes to control verification](#sec-command-line-options-and-attributes-for-verification).

### 24.8.1. Verification debugging {#sec-verification-debugging}

Let's assume one assertion is failing ("assertion might not hold" or "postcondition might not hold"). What should you do next?

The following section is textual description of the animation below, which illustrates the principle of debugging an assertion by computing the weakest precondition:  

> ![weakestpreconditionDemo](https://user-images.githubusercontent.com/3601079/157976402-83fe4d37-8042-40fc-940f-bcfc235c7d2b.gif)

Author: Agreed. This one was shot in 2 minutes. We will dedicate a better resource later once someone finds the courage to do so with voiceover.

  if b {
    i := j;
    return;
  }//^^^^^^^ a postcondition might not hold on this return path.
  i := 2;
}
```
Now, you can assert the postcondition just before the return:
```dafny
method FailingPostcondition(b: bool) returns (i: int)
  ensures 2 <= i
{
  var j := if !b then 3 else 1;
  if b {
    i := j;
    assert 2 <= i; // This assertion might not hold
    return;
  }
  i := 2;
}
```
That's it! Now the postcondition is not failing anymore, but the `assert` contains the error!
you can now move to the next section to find out how to debug this `assert`.

#### 24.8.1.2. Failing asserts {#sec-failing-asserts}
In the [previous section](#sec-failing-postconditions), we arrive to the point where we have a failing assertion:
```dafny
method FailingPostcondition(b: bool) returns (i: int)
  ensures 2 <= i
{
  var j := if !b then 3 else 1;
  if b {
    i := j;
    assert 2 <= i; // This assertion might not hold
    return;
  }
  i := 2;
}
```
To debug why this assert might not hold, we need to _move this assert up_, which is similar to [_computing the weakest precondition_](https://en.wikipedia.org/wiki/Predicate_transformer_semantics#Weakest_preconditions).

> For example, if we have `x := Y; assert F(x);` and the `assert F(x);` might not hold, the weakest precondition for it to hold before `X := Y;` can be written as the assertion `assert F(Y);`, where we replace every occurence of `x` in `F(x)` into `Y`.

Reviewer 2: I take it back, since `x` is used like a variable declaration lowercase is more idiomatic.

Now, either the error is obvious, or we can one more time replace `j` by its value and create the assert `assert b ==> ((if !b then 3 else 1) >= 2);`
```dafny
method FailingPostcondition(b: bool) returns (i: int)
  ensures 2 <= i
{
  assert b  ==>  2 <= (if !b then 3 else 1);  // This assertion might not hold
  var j := if !b then 3 else 1;
  assert b  ==>  2 <= j;
  if b {
    assert 2 <= j;
    i := j;
    assert 2 <= i;
    return;
  }
  i := 2;
}
```
At this point, this is pure logic. We can simplify the assumption:
```
b ==>  2 <= (if !b then 3 else 1)
!b ||  (if !b then 2 <= 3 else 2 <= 1)
!b ||  (if !b then true else false)
!b || !b;
!b;
```
Now we can understand what went wrong: When b is true, all of these formulas above are false, this is why the Dafny verifier was not able to prove them.
In the next section, we will explain how to `move asserts up` in certain useful patterns.

#### 24.8.1.3. Failing asserts cases {#sec-failing-asserts-special-cases}

This list is not exhaustive but can definitely be useful to provide the next step to figure out why Dafny could not prove an assertion.

>  Failing assert           | Suggested rewriting

Reviewer 2: I would format all of the examples in the table with \`\`\`dafny ... \`\`\` code blocks instead of just \` ... \` on each line. I think:
```dafny
assert B ==> P;
if B {
  assert P;
  ...
}
```

Is a lot nicer looking than:

`assert B ==> P;`<br>`if B {`<br>&nbsp;&nbsp;`  assert P;`<br>&nbsp;&nbsp;`  ...`<br>`}`

Or is it not possible to put code blocks inside a table cell?

Author: As far as I tested and tested and tested, It is not possible to put code blocks inside a table cell. I have tried many times without success.

Reviewer 2: Your effort is appreciated at least! This is fine for now but let's cut a follow-up issue - I think it would be worth reformatting these examples without a table to make them easier to read and understand.

--------------------------|---------------------------------------

>  <br>`x := Y;`<br>`assert P[x];` | `assert P[Y];`<br>`x := Y;`<br>`assert P[x];`

Reviewer 2: I would define the `P[x]` notation first, since you didn't use it earlier (and perhaps doing so in earlier sections where you say things like "replace every occurence of")

 <br>`if B {`<br>&nbsp;&nbsp;`  assert P;`<br>&nbsp;&nbsp;`  ...`<br>`}` | `assert B ==> P;`<br>`if B {`<br>&nbsp;&nbsp;`  assert P;`<br>&nbsp;&nbsp;`  ...`<br>`}`
 <br>`if B {`<br>&nbsp;&nbsp;`  ...`<br>`} else {`<br>&nbsp;&nbsp;`  assert P;`<br>&nbsp;&nbsp;`  ...`<br>`}` | `assert !B ==> P;`<br>`if B {`<br>&nbsp;&nbsp;`  ...`<br>`} else {`<br>&nbsp;&nbsp;`  assert P;`<br>&nbsp;&nbsp;`  ...`<br>`}`
 <br><br>`if X {`<br>&nbsp;&nbsp;`  ...`<br>`} else {`<br>&nbsp;&nbsp;`  ...`<br>`}`<br>`assert A;` | `if X {`<br>&nbsp;&nbsp;`  ...`<br>&nbsp;&nbsp;`  assert A;`<br>`} else {`<br>&nbsp;&nbsp;`  ...`<br>&nbsp;&nbsp;`  assert A;`<br>`}`<br>`assert A;`
 <br><br><br><br><br>`assert forall x :: Q(x);` | [`forall x`](#sec-forall-statement)<br>&nbsp;&nbsp;`  ensures Q(x)`<br>`{`<br>&nbsp;&nbsp;`  assert Q(x);`<br>`};`<br>` assert forall x :: Q(x);`
 <br><br><br><br><br>`assert forall x :: P(x) ==> Q(x);` | [`forall x | P(x)`](#sec-forall-statement)<br>&nbsp;&nbsp;`  ensures Q(x)`<br>`{`<br>&nbsp;&nbsp;`  assert Q(x);`<br>`};`<br>` assert forall x :: P(x) ==> Q(x);`
 <br>`assert exists x :: P(x);`<br> | `assert P(x0);`<br>`assert exists x :: P(x);`<br>for a given expression `x0`.

>  <br>`ensures exists i :: P(i);`<br> | `returns (j: int)`<br>`ensures P(j) ensures exists i :: P(i)`<br>in a lemma, so that the `j` can be computed explicitely.

Reviewer 2: ```suggestion
 <br>`ensures exists i :: P(i);`<br> | `returns (j: int)`<br>`ensures P(j) ensures exists i :: P(i)`<br>in a lemma, so that the `j` can be computed explicitly.
```

  <br><br>`assert forall i | i == m :: P(m);` |  `assert P(m);`<br>`assert forall i | i == m :: P(i);`
  `method m(i) returns (j: T)`<br>&nbsp;&nbsp;`  requires A(i)`<br>&nbsp;&nbsp;`  ensures B(i, j)`<br>`{`<br>&nbsp;&nbsp;`  ...`<br>`}`<br><br>`method n() {`<br>&nbsp;&nbsp;`  ...`<br><br><br>&nbsp;&nbsp;`  var x := m(a);`<br>&nbsp;&nbsp;`  assert P(x);` | `method m(i) returns (j: T)`<br>&nbsp;&nbsp;`  requires A(i)`<br>&nbsp;&nbsp;`  ensures B(i, j)`<br>`{`<br>&nbsp;&nbsp;`  ...`<br>`}`<br><br>`method n() {`<br>&nbsp;&nbsp;`  ...`<br>&nbsp;&nbsp;`  assert A(k);`<br>&nbsp;&nbsp;`  assert forall x :: B(k, x) ==> P(x);`<br>&nbsp;&nbsp;`  var x := m(k);`<br>&nbsp;&nbsp;`  assert P(x);`

### 24.8.2. Assertion batches {#sec-assertion-batches}

To understand how to control verification,
it is first useful to understand how Dafny verifies functions and methods.

For every method (or function, constructor, etc.), Dafny extracts _assertions_, for example:

* any explicit [`assert` statement](#sec-assert-statement) is _an assertion_[^precision-requires-clause].

> * A consecutive pair of lines in a [`calc` statement](#sec-calc-statement) forms _an assertion_.

Reviewer 2: Consider adding "...that the expressions are related according to the common operator"

> * Every function or method call with a [`requires` clause](#sec-requires-clause) yields _one assertion per requires clause_[^precision-requires-clause]

Reviewer 2: ```suggestion
* Every call to a function or method with a [`requires` clause](#sec-requires-clause) yields _one assertion per requires clause_[^precision-requires-clause]
```

  (special cases such as sequence indexing come with a special require clause that the index is within bounds).
* Assignments `o.f := E;` yield an _assertion_ that `o.f` is allowed by the enclosing [`modifies` clause](#sec-loop-framing).
* [Assign-such-that operators](#sec-update-and-call-statement) `x :| P(x)` yield an _assertion_ that `exists x :: P(x)`.
* Every [`ensures` clause](#sec-ensures-clause) yields an _assertion_ at the end of the method and on every return.
* Every [`decreases` clause](#sec-decreases-clause) yields an _assertion_ at either a call site or at the end of a while loop.
* Every non-empty [subset type](#sec-subset-types) yields an _assertion_ that its witness satisfies the constraint.

[^precision-requires-clause]: Dafny actually breaks things down further. For example, a precondition `requires A && B` or an assert statement `assert A && B;` turns into two assertions, more or less like `requires A requires B` and `assert A; assert B;`.

> It is useful to mentally visualize all these assertions as a list (an _assertion batch_) that roughly follows the order in the code[^complexity-path-encoding],
> except for `ensures` assertions that appear earlier in the code but, for verification purposes, would appear at the end.
> Thus, to prove or disprove a single assertion within this list, all the assumptions Dafny needs are 1) the global context,
> and 2) every preceding assumption (and previous assertions transformed in assumptions).

Reviewer 2: This doesn't QUITE nail it for me, as I think it introduces "assertion batch" slightly too early. How about:

```suggestion
Each assertion is not just a boolean expression that must be true, but also depends on context, such as the definition of variables and other assertions that appear earlier in the control flow[^complexity-path-encoding]. The fundamental unit of verification in Dafny is an _assertion batch_: one or more assertions along with all the context they depend on. To reduce overhead, by default Dafny collects all the assertions in the body of a given method into a single batch that it sends to the verifier, which tries to prove it correct.
```

Author: I am doing a change that merges both what I did and what you suggest. Let me know if you like it.

[^complexity-path-encoding]: All the complexities of the execution paths (if-then-else, loops, goto, break....) are, down the road and for verification purposes, cleverly encoded with variables recording the paths and guarding assumptions made on each path. In practice, a second clever encoding of variables enables grouping many assertions together, and recovers which assertion is failing based on the value of variables that the SMT solver returns.

To reduce overhead, Dafny collects all the assertions of one method into one single conjecture that it sends to the verifier, which tries to prove it correct:

* If the verifier says it is correct[^smt-encoding], it means that all the assertions hold.
* If the verifier returns a counterexample, this counter-example is used to determine both the failing assertion and the failing path.

>   In order to retrieve additional failing assertions, Dafny will again query the verifier after turning previously failed assertions into assumptions, [like this](https://github.com/dafny-lang/dafny/discussions/1898#discussioncomment-2344533).[^caveat-about-assertion-and-assumption]

Reviewer 2: I wouldn't link to a discussion from here - there's no telling what content other users will add there in the future and be implicitly endorsed as correct by this link. 

In any case this feels like detail we don't need at this point. The bullet points are all implementation detail and can be moved to footnotes or elsewhere. The important point is just that an assertion batch is verified all at once and produces a single result. See my proposed edit above.

Author: I am linking the post, so let me 1) make it explicitly known that I link to the post and not the discussion and 2) move this into a footnote, which have less endorsement than a link.

> * If the verifier returns `unknown` or times out, or even preemptively for difficult assertions or to reduce the chance that the verifier will ‘be confused’ by the many assertions in a large batch, Dafny partitions the assertions up into smaller batches[^smaller-batches]. An extreme case is the use of the `/vcsSplitOnEveryAssert` command-line option or the [`{:vcs_split_on_every_assert}` attribute](#sec-vcs_split_on_every_assert), which causes Dafny to make one batch for each assertion.

Reviewer 2: Just for my own sanity: I know Boogie has options such as `/vcsMaxKeepGoingSplits` that cause this adaptive behavior, but my understanding is that they are not on by default, and that is critical to make sure verification is deterministic.

Author: Correct.

Reviewer 2: Okay, then we need to clarify that this bullet isn't true by default.

[^caveat-about-assertion-and-assumption]: Caveat about assertion and assumption: One big difference between an "assertion transformed in an assumption" and the original "assertion" is that the original "assertion" can unroll functions twice, whereas the "assumed assertion" can unroll them only once. Hence, Dafny can still continue to analyze assertions after a failing assertion without automatically proving "false" (which would make all further assertions vacuous).

[^smaller-batches]: To create a smaller batch, Dafny duplicates the assertion batch, and transforms each assertion into an assumption into exactly one batch, so that assertions are verified only in one batch. This results in "easier" formulas for the verifier because it has less to prove, but it takes more overhead because every verification instance have a common set of axioms and there is no knowledge sharing between instances because they run independently.

### 24.8.3. Controlling assertion batches {#sec-assertion-batches-control}

> Here is how you can control how Dafny partition assertions into batches.

Reviewer 2: ```suggestion
Here is how you can control how Dafny partitions assertions into batches.
```

* [`{:focus}`](#sec-focus) on an assert generates a separate assertion batch for the assertions of the enclosing block.
* [`{:split_here}`](#sec-split_here) on an assert generates a separate assertion batch for assertions after this point.

> * [`{:vcs_split_on_every_assert}`](#sec-vcs_split_on_every_assert) on a function or a method generates one assertion batch per assertion

Reviewer 2: This is awesome - with all the background above these become much more understandable, and easy to describe with a single sentence each!

You can also provide _heuristics_ for the verifier to split assertion batches, if you are not using the fine-grained attributes above. The effect of these attributes may vary, because they are low-level attributes and tune low-level heuristics.
* [`{:vcs_max_cost N}`](#sec-vcs_max_cost) on a function or method enables splitting the assertion batch until the "cost" of each batch is below N.
  Usually, you would set [`{:vcs_max_cost 0}`](#sec-vcs_max_cost) and [`{:vcs_max_splits N}`](#sec-vcs_max_splits) to ensure it generates N assertion batches.

> * [`{:vcs_max_keep_going_splits N}`](#sec-vcs_max_keep_going_splits) where N > 1 on a method dynamically splits the initial assertion batch up to N components if the verifier is stuck the first time.

Reviewer 2: As per my comment above, I wouldn't encourage Dafny users to use these attributes given the non-deterministic results, and the total ambiguity of what "cost" means here.

Author: "We discourage the use of the following _heuristics attributes_ to partition assertions into batches.
The effect of these attributes may vary, because they are low-level attributes and tune low-level heuristics, and will result in splits that could be manually controlled anyway."

> ### 24.8.4. Command-line options and other attributes to control verification {#sec-command-line-options-and-attributes-for-verification}

Reviewer 2: This section would be a great spot to mention the `dafny-reportgenerator` too, although that can wait for the next PR (since we'd want some background content on verification stability as well to motivate it)

There are many great options that control various aspects of verifying dafny programs. Here we mention only a few:

- Control of output: [`/dprint`](#sec-controlling-output), [`/rprint`](#sec-controlling-output), `/stats`, [`/compileVerbose`](#sec-controlling-compilation)
- Whether to print warnings: `/proverWarnings`

> - Control of time: `/timeLimit`

Reviewer 2: Definitely add `/rlimit` and `{:rlimit}` here too!

Author: I added them here but we will need a separate PR to document them, as this PR is almost at the end of its life :-)

[^precision-requires-clause]: Dafny actually breaks things down further. For example, a precondition `requires A && B` or an assert statement `assert A && B;` turns into two assertions, more or less like `requires A requires B` and `assert A; assert B;`.

It is useful to mentally visualize all these assertions as a list that roughly follows the order in the code,
except for `ensures` or `decreases` that generate assertions that seem earlier in the code but, for verification purposes, would be placed later.
In this list, each assertion depends on other assertions, statements and expressions that appear earlier in the control flow[^complexity-path-encoding].

[^complexity-path-encoding]: All the complexities of the execution paths (if-then-else, loops, goto, break....) are, down the road and for verification purposes, cleverly encoded with variables recording the paths and guarding assumptions made on each path. In practice, a second clever encoding of variables enables grouping many assertions together, and recovers which assertion is failing based on the value of variables that the SMT solver returns.

> The fundamental unit of verification in Dafny is an _assertion batch_, which consists of one or more assertions from this "list", along with all the remaining assertions turned into assumptions. To reduce overhead, by default Dafny collects all the assertions in the body of a given method into a single assertion batch that it sends to the verifier, which tried to prove it correct.

Reviewer 2: ```suggestion
The fundamental unit of verification in Dafny is an _assertion batch_, which consists of one or more assertions from this "list", along with all the remaining assertions turned into assumptions. To reduce overhead, by default Dafny collects all the assertions in the body of a given method into a single assertion batch that it sends to the verifier, which tries to prove it correct.
```

* If the verifier says it is correct[^smt-encoding], it means that all the assertions hold.
* If the verifier returns a counterexample, this counterexample is used to determine both the failing assertion and the failing path.
  In order to retrieve additional failing assertions, Dafny will again query the verifier after turning previously failed assertions into assumptions.[^example-assertion-turned-into-assumption] [^caveat-about-assertion-and-assumption]

> * If the verifier returns `unknown` or times out, or even preemptively for difficult assertions or to reduce the chance that the verifier will ‘be confused’ by the many assertions in a large batch, Dafny partitions the assertions up into smaller batches[^smaller-batches]. An extreme case is the use of the `/vcsSplitOnEveryAssert` command-line option or the [`{:vcs_split_on_every_assert}` attribute](#sec-vcs_split_on_every_assert), which causes Dafny to make one batch for each assertion.

Reviewer 2: ```suggestion
* If the verifier returns `unknown` or times out, or even preemptively for difficult assertions or to reduce the chance that the verifier will ‘be confused’ by the many assertions in a large batch, Dafny may partition the assertions into smaller batches[^smaller-batches]. An extreme case is the use of the `/vcsSplitOnEveryAssert` command-line option or the [`{:vcs_split_on_every_assert}` attribute](#sec-vcs_split_on_every_assert), which causes Dafny to make one batch for each assertion.
```
